pennylane/labs/resource_estimation/templates/comparators.py

```diff
@@ -0,0 +1,543 @@
+
+    @property
+    def resource_params(self):
+        r"""Returns a dictionary containing the minimal information needed to compute the resources.
+
+        Returns:
+            dict: A dictionary containing the resource parameters:
+                * a_num_qubits (int): the size of the first register
+                * b_num_qubits (int): the size of the second register
+                * geq (bool): If set to ``True``, the comparison made will be :math:`a \geq b`. If
+                  ``False``, the comparison made will be :math:`a < b`.
+
+        """
+        return {
+            "a_num_qubits": self.a_num_qubits,
+            "b_num_qubits": self.b_num_qubits,
+            "geq": self.geq,
+        }
+
+    @classmethod
+    def resource_rep(cls, a_num_qubits, b_num_qubits, geq=False):
+        r"""Returns a compressed representation containing only the parameters of
+        the Operator that are needed to compute the resources.
+
+        Args:
+            a_num_qubits (int): the size of the first register
+            b_num_qubits (int): the size of the second register
+            geq (bool): If set to ``True``, the comparison made will be :math:`a \geq b`. If
+                ``False``, the comparison made will be :math:`a < b`.
+
+        Returns:
+            CompressedResourceOp: the operator in a compressed representation
+        """
+        return CompressedResourceOp(
+            cls, {"a_num_qubits": a_num_qubits, "b_num_qubits": b_num_qubits, "geq": geq}
+        )
+
+    @classmethod
+    def default_resource_decomp(cls, a_num_qubits, b_num_qubits, geq=False, **kwargs):
+        r"""Returns a dictionary representing the resources of the operator. The
+        keys are the operators and the associated values are the counts.
+
+        Args:
+            a_num_qubits (int): the size of the first register
+            b_num_qubits (int): the size of the second register
+            geq (bool): If set to ``True``, the comparison made will be :math:`a \geq b`. If
+                ``False``, the comparison made will be :math:`a < b`.
+
+        Resources:
+            The resources are obtained from appendix B, Figure 3 in `arXiv:1711.10460
+            <https://arxiv.org/abs/1711.10460>`_ for registers of same size.
+            If the size of registers differ, unary iteration technique from
+            `Babbush et al. (2018) <https://arxiv.org/pdf/1805.03662>`_ is used
+            to combine the results from extra qubits.
+
+        Returns:
+            list[GateCount]: A list of GateCount objects, where each object
+            represents a specific quantum gate and the number of times it appears
+            in the decomposition.
+        """
+
+        gate_list = []
+        compare_register = min(a_num_qubits, b_num_qubits)
+
+        one_qubit_compare = resource_rep(plre.ResourceSingleQubitCompare)
+        two_qubit_compare = resource_rep(plre.ResourceTwoQubitCompare)
+        if a_num_qubits == b_num_qubits:
```

Comment:
What happens if the size of the two registers being compared is both one or both two? I think we should return just a single instance of SingleQubitCompare, or TwoQubitCompare in this "base" case.